File: Java/lab16_Iterator/task_2_1 – solution/src/IntegerIterator.java

import java.util.ArrayList;
import java.util.Iterator;

public class IntegerIterator implements Iterator<Integer> {

    final private ArrayList<Integer> collectionIntegers  = new ArrayList<>();
    private int index = 0;

    public IntegerIterator(CollectionIntegers collection) {
        collectionIntegers.addAll(collection.getArray());
        collectionIntegers.sort(Integer::compareTo);
    }

    @Override
    public boolean hasNext() {
        return index < collectionIntegers.size();
    }

    @Override
    public Integer next() {
        return collectionIntegers.get(index++);
    }
}

File: Cpp/lab16_Iterator/task_2_1 - solution/src/integer_iterator.cpp
#include <algorithm>

#include "integer_iterator.hpp"
#include "collection_integer.hpp"


IntegerIterator::IntegerIterator(CollectionIntegers const* collection)
: collectionIntegers(collection->getArray()), index(0)
{
  std::sort(this->collectionIntegers.begin(), this->collectionIntegers.end());
}

bool IntegerIterator::hasNext() const
{
  return this->index < this->collectionIntegers.size();
}

int IntegerIterator::next()
{
  return this->collectionIntegers.at((this->index)++);
}

int IntegerIterator::operator++(int)
{
  return this->next();
}
